### DISCUSSION TITLE
How to handle ref invalidation during async gaps?

### DISCUSSION THREAD

**Discussion by btrautmann** [rrousselGit/riverpod#3177] [General]
Hello 👋 

I'm running into an issue that I've seen mentioned only a couple times, where I receive the following error between async gaps in `Provider`s, specifically those that are `await`ing another `Provider`'s `Future` or `Stream`.

```console
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 'package:riverpod/src/framework/element.dart': Failed assertion: line 673 pos 7: '!_didChangeDependency': Cannot use ref functions after the dependency of a provider changed but before the provider rebuilt
```
I understand what the error is telling me, and you highlight that in your comment [here](https://github.com/rrousselGit/riverpod/issues/815#issuecomment-950135064), but this happens in cases where there are no `await`s in the `Provider` "chain" other than those `await`ing other `Provider`s. 

One such case is below:
```dart
// Both categoryGroups and tracker await FutureProviders that await StreamProviders that yield models from an `Isar` database
final categoryGroups = await ref.watch(categoryGroupsProvider.future);
final tracker = await ref.watch(spendTrackerProvider(trackerId).future);
final incomeFilter = ref.watch(incomeFilterProvider); // Error occurs here
final groupByDate = ref.watch(groupByDateProvider);
```

If, during these async gaps, the `ref` is being invalidated, how am I supposed to guard against that? I don't see an API similar to `mounted` for `StateNotifier`.

As a follow-up, am I abusing `ref.watch(provider.future)`? Is this an anti-pattern, or is this a pretty expected use case? It seems that these async gaps would be pretty commonplace in `Provider` code utilizing `Future`/`StreamProvider`, but I'm unsure if I'm just doing something dumb.

Thanks for any help!

**Comment by rrousselGit** [comment]
> If, during these async gaps, the ref is being invalidated, how am I supposed to guard against that? I don't see an API similar to mounted for StateNotifier.

Use ref.onDispose to detect if the provider was disposed during an await.  That generally should help
